Add hint-based footnote extension

moben opened the pull request:
# What / Why

As promised (threatened?) in #617, this adds support for in-page footnotes that span more than one node.

This adds a hint-based way to allow marking nodes that continue an in-page footnote that starts in a previous node. This is necessary to make in-page footnotes usable in many books with footnotes like this:

```
<p class="footnote" id="f1">footnote 1 starts here</p>
<p>footnote 1 continues here</p>
<p class="footnote" id="f2">footnote 2 starts here</p>
```

(one could of course argue that the publisher did a poor job)
# How

This introduces a new hint to mark such footnotes, as well as a way to target this hint more accurately. These are split across two commits that work independently.

## `-cr-hint: extend-footnote-inpage`

If there is a node with `-cr-hint: footnote-inpage` immediately followed by a node with `-cr-hint: extend-footnote-inpage` then this new hint causes the nodes to act like one joined footnote block. Specifically:

- Append lines from the `extend-footnote-inpage` block to the lines in the `actual` footnote of the first block.
- Accumulate anchor `id`s for targeting the footnote across both blocks
 
It's allowed to chain this across multiple consecutive blocks with `-cr-hint: extend-footnote-inpage`.

## `-cr-only-if: following-inpage-footnote`

In theory `-cr-hint: extend-footnote-inpage` is enough to extend footnotes using e.g. the `~` sibling selector if one is willing to write enough css (and if `crengine` supports the necessary selectors). For example to allow footnote extension for all footnotes matched by koreader's _In-page classic classname footnotes_, one can write

```
.footnote ~ *, .footnotes ~ *, .fn ~ *,
.note ~ *, .note1 ~ *, .note2 ~ *, .note3 ~ *,
.ntb ~ *, .ntb-txt ~ *, .ntb-txt-j ~ *,
.przypis ~ *, .przypis1 ~ *,
.voetnoten ~ *
{
    -cr-only-if: -inpage-footnote;
    -cr-hint: extend-footnote-inpage;
}

.footnote ~ autoBoxing, .footnotes ~ autoBoxing, .fn ~ autoBoxing,
.note ~ autoBoxing, .note1 ~ autoBoxing, .note2 ~ autoBoxing, .note3 ~ autoBoxing,
.ntb ~ autoBoxing, .ntb-txt ~ autoBoxing, .ntb-txt-j ~ autoBoxing,
.przypis ~ autoBoxing, .przypis1 ~ autoBoxing,
.voetnoten ~ autoBoxing
{
    -cr-only-if: -inpage-footnote;
    -cr-hint: extend-footnote-inpage;
}
```

To extend all footnotes as far as possible (assuming there are no non-footnote nodes between them)

But this is quite repetitive and does not allow writing a generic style tweak independent of the specific footnote class. It would be nice to write something like this (pseudocode)

```
*[-cr-only-if: inpage-footnote] ~ *, *[-cr-only-if: inpage-footnote] ~ autoBoxing {
    -cr-only-if: -inpage-footnote;
    -cr-hint: extend-footnote-inpage;
}
```

This adds implements something close to these semantics as `-cr-only-if: following-inpage-footnote`, which matches if the previous sibling is an inpage footnote (potentially extended as per the rules above).

That allows writing the above much simpler like this (which I would propose as a footnote style tweak in `koreader` if this lands in `crengine`):

```
*, autoBoxing {
    -cr-only-if: following-inpage-footnote -inpage-footnote;
    -cr-hint: extend-footnote-inpage;
}
```

If new rules are added to set `-cr-hint: footnote-inpage` on other nodes, the rule to extend them does not need to be changed.

The simplification is even better if the selector is more complicated, e.g. if the footnotes are separated by headings:

```
*:not(h1, h2, h3, h4, h5, h6), autoBoxing{
    -cr-only-if: following-inpage-footnote -inpage-footnote;
    -cr-hint: extend-footnote-inpage;
}
```

This might be impossible to express in vanilla css selectors because unlike `~`, it will stop at the heading and not continue afterwards. That is:

```
<p class="footnote" />  ← footnote
<p />                   ← above hint applies
<h1 />                  ← hint doesn't apply
<p />                   ← hint dosen't apply, but `~` would
```

In the books that I have been able to test with, this matches what one wants. _"Footnotes are continued until element X or class Y_

## Why Both

You might think that `-cr-only-if: following-inpage-footnote` might be enough because one could write

```
*, autoBoxing {
    -cr-only-if: following-inpage-footnote -inpage-footnote;
    -cr-hint: footnote-inpage;
}
```

Indeed, this works in simple cases but breaks as soon as the extended footnote (recursively) contains another `id` attribute. Then its lines will be added to a separate `LVFootNote`. This will happen if the footnote contains nested footnote links, as these usually have `id`s for back-links. (I've seen this in my copy of _Infinite Jest_)

On the other hand, just having `-cr-hint: extend-footnote-inpage` works, but writing style tweaks is less ergonomic. Being able to use `*:not(...)` with `-cr-only-if: following-inpage-footnote` is much less error prone to exclude unwanted nodes.

# Test Case

I've been testing with a few different books that are heavy on footnotes and even nested footnotes (with #617 also applied). Especially _Infinite Jest_ and _Jonathan Strange & Mr Norrell_. I've also made those even worse in an attempt to break this code and find edge cases, but even something like e.g. the following works:

```
<div class="footnote" id="endnote_text_24">
	<p>
		<a href="chap-8.xhtml#endnote_reference_24">
			24.
		</a>
		first text with sub-footnote reference
		<sup>
			<a href="note-1a.xhtml#footnote_number_4" id="footnote_reference_4">
				a
			</a>
		</sup>
	</p>
</div>
<p class="more-footnote">
	extended footnote text
</p>
free-floating footnote text in autoBox
<p class="more-footnote">
	more of the same footnote with another sub-footnote
	<sup>
		<a href="note-1a.xhtml#footnote_number_5" id="footnote_reference_5">
			b
		</a>
	</sup>
</p>
<p class="more-footnote">
	even more extended footnote text
</p>
<p class="footnote" id="footnote_number_4">
	sub-footnote text
</p>
<!-- more sub footnotes and normal footnotes -->
```

With the style tweak suggested above, this produces a footnote along the lines of

> 24. first text with sub-footnote referenceᵃ
>   extended footnote text
>   free-floating footnote text in autoBox
>   more of the same footnote with another sub-footnoteᵇ
>   even more extended footnote text
> ... nested footnotes follow

(unless I made a mistake in editing my example for this description)

# Implementation notes

Does *not* depend on #617 (though I also tested them together).

I originally tried to keep track if a given block can extend a previous footnote inside `lvrend` / `renderBlockElementEnhanced`, but that turned out quit error prone due to its recursive nature. So I moved to tracking this in `LVRendPageContext`. I believe this is much more robust and the logic of _"if no regular lines have been added the footnote can still be extended"_ makes intuitive sense and makes the change quite small. But I'd be open to suggestions if there's a better way.

# Aside: Previous Attempts

As originally threatened in #617, I did also attempt to achieve extending footnotes by re-using the heuristic in `koreader-base` that is used in popup footnotes.

I extended `ldomDocument` to allow passing an `ldomCallback` that detects footnotes and stores them as ranges in `ldomDocument`. This callback wound be invoked after parsing but before rendering.

While this works it does not really fit with `crengine` design and has big risks of regressing user experience for books with exceptionally bad markup. It would also need a lot of knobs in `koreader` to adjust the footnote detection.

I also hit some insurmountable problems with this approach because you'd want to use `-cr-hint`'s as inputs to the footnote heuristic to mark footnotes.  But you also want to style nodes after they were detected as footnotes.

If you're curious, that attempt is at:

- https://github.com/moben/crengine/commits/footnotes_heuristic
- https://github.com/moben/koreader-base/commits/footnotes_heuristic
- https://github.com/moben/koreader/commits/footnotes_heuristic

Thanks @poire-z for the thoughts and links in https://github.com/koreader/crengine/pull/617#issuecomment-2676469220, that led me to exploring this hint-based solution and I think it's much better :)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/crengine/618)
<!-- Reviewable:end -->

Frenzie commented:
> (one could of course argue that the publisher did a poor job)

I don't think one could. Poor is much too kind.

moben commented:
> I don't think one could. Poor is much too kind.

Sure. On the other hand, most of my books that have footnotes look like this :confused:

moben commented:
Demo style tweaks with / without `-cr-only-if: following-inpage-footnote`: https://github.com/koreader/koreader/pull/13554

Frenzie commented:
Anyway, nice work. I'll leave the review to @poire-z but it looks clean to me.

poire-z commented on crengine/src/lvrend.cpp:
```diff
@@ -7278,14 +7278,25 @@ void renderBlockElementEnhanced( FlowState * flow, ldomNode * enode, int x, int
+        if ( STYLE_HAS_CR_HINT(style, FOOTNOTE_INPAGE) ) {
+            enode->getAllInnerAttributeValues(attr_id, footnoteIds);
+        } else if ( STYLE_HAS_CR_HINT(style, EXTEND_FOOTNOTE_INPAGE) ) {
+            if (lString32 last_footnoteid = flow->getPageContext()->getCurrentFootNoteId();
+                 ! last_footnoteid.empty()) {
```
(Not fond of this syntax - creating variables in the `if ( )` condition. If there's no other more simple way, may be put it all on a single line, so it's easier to grasp it is all in the `( )`.)

poire-z commented on crengine/src/lvstsheet.cpp:
```diff
@@ -3416,6 +3416,7 @@ bool LVCssDeclaration::parse( const char * &decl, bool higher_importance, lxmlDo
                             hints = CSS_CR_HINT_NONE_NO_INHERIT;
                         }
                         else if ( substr_icompare("footnote-inpage", decl) )        hints |= CSS_CR_HINT_FOOTNOTE_INPAGE|CSS_CR_HINT_INSIDE_FOOTNOTE_INPAGE;
+                        else if ( substr_icompare("extend-footnote-inpage", decl) ) hints |= CSS_CR_HINT_EXTEND_FOOTNOTE_INPAGE|CSS_CR_HINT_INSIDE_FOOTNOTE_INPAGE;
```
Just asking: is it ok to set and carry `INSIDE_FOOTNOTE_INPAGE`, even if the `extend-footnote-inpage` matching node does not end up extending a footnote inpage ?
We use it to fixup styling, so we may then end up fixing stuff that is not a inpage footnote.
If it is not ok, dunno if it's better to unhint it in lvrend when not `appendingFootnote` , or to hint it there when 
`appendingFootnote`.

poire-z replied:
Mhhh, it might be too late to update the hint (needs to use copystyle()), as it is inherited, so you would need to also update all the children that have inherited it...
And it not updating it, but just handling this when checking in LVCssDeclaration::apply(), it might be costly.
No obvious idea :/

moben replied:
I was thinking about this but decided against it because it got quite complicated and `footnote-inpage` already works the same. Even if something isn't actually referenced as a footnote it's still styled like one.

Also, updating it later would require re-computing the style of the node and all children due to `-cr-only-if: inside-inpage-footnote`.

The reason I didn't do it in `apply` is that I added `LVRendPageContext::getCurrentFootNoteId()` to track if extending a footnote is allowed or not. This is quite flexible because it just cares if there were any non-footnote lines and avoids having to walk the preceding nodes in `lvrend` and check a lot of edge cases (e.g. if the preceding footnote was actually `display: inline` or `display: none`).

To do it in `LVCssDeclaration::apply()` I'd have to very strictly define when something can / cannot be an (extended) footnote in terms of nodes and styles.

poire-z reviewed (commented):
Reviewing first commit:
I'm fine with the idea, it's simple and intuitive (and didn't even need the essay in the PR :))
It just wastes a hint bit, but ok, it's worth it, I would have been happy having it a few times in the past.
I'm just no longer really familiar with the lvpagesplitter footnote handling (and couldn't really get familiar looking at it all for 20 mn), so trusting you. (And I guess - please recheck and confirm this - that there is no behavioral change as long as we don't use that new hint).

poire-z reviewed (commented):
Reviewing second commit:
Fine with it, it costs nothing (-cr-only-if don't use a precious bit).

moben commented on crengine/include/cssdef.h:
```diff
@@ -464,6 +464,7 @@ enum css_generic_value_t {
                                                        //                           footnote block container)
 #define CSS_CR_HINT_FOOTNOTE_IGNORE         0x08000000 // -cr-hint: footnote-ignore block is not a footnote (even if
                                                        //                           everything else indicates it is)
+#define CSS_CR_HINT_EXTEND_FOOTNOTE_INPAGE  0x10000000 // -cr-hint: extend-footnote-inpage
```
I like `extend-footnote-inpage` a bit better because it's not a different variation of `footnote-inpage` but something in addition to it.

Also, I had `extending` first and it's too long so we'd have to re-align all the defines and some code in `lvstsheet.cpp` :upside_down_face:

moben commented on crengine/include/lvpagesplitter.h:
```diff
@@ -443,7 +444,7 @@ class LVRendPageContext
+                if ( ref.get()->isActual() && ! appending ) {
                     // If the one we found is already actual, something is wrong: this may
                     // happen with buggy books with duplicated id= (ie. Wikipedia EPUBs...).
                     // LVPageSplitter expects a footnote to be a single chunk/slice of the
```
Extended a comment to hopefully make it a bit clearer.

The function will always create a footnote if it doesn't exist, but `appending` (renamed it to `allow_appending` just now) is only checked in the case where we find a pre-existing footnote. It's basically needed to distinguish between the case of a broken book that has multiple nodes with the same `id` and us wanting to add more lines to an existing footnote.

I think lvrend is the only place where that can be known. If we'd just check here if `getCurrentFootNoteId() == id` then' we'd append the lines of two consecutive nodes with duplicate `id` instead of having the second footnote overwrite the first.

poire-z replied:
Btw, how would it work if we set (or more correctly: if it matches) `extend-footnote-inpage` not on a sibling, but on a child or a sibling child or on a parent sibling ?

moben replied:
### child

Should behave the same as if you set `footnote-inpage` on a child of `footnote-inpage`. You get a `CRLog::error("Nested entering note" )` from `enterFootNote` and it becomes a no-op so the child lines are part of the same footnote. But once the child calls `leaveFootNote` other children will be incorrectly excluded from the footnote.

### sibling child / parent sibling

Depends on if there are lines before the extended footnote sibling child / parent sibling (e.g. text nodes that aren't `extend-footnote-inpage`). If no lines between them then they are part of the footnote, otherwise not. Essentially anything that's contiguous in terms of `LVRendPageContext::AddLine`, is `extend-footnote-inpage` and meets normal footnote requirements (I think just `display: block`?) is part of the footnote.

moben commented on crengine/src/lvstsheet.cpp:
```diff
@@ -3113,6 +3114,7 @@ enum cr_only_if_t {
     cr_only_if_not_inpage_footnote,
     cr_only_if_inside_inpage_footnote,
     cr_only_if_not_inside_inpage_footnote,
+    cr_only_if_following_inpage_footnote,
     cr_only_if_line_height_normal,
     cr_only_if_not_line_height_normal,
```
I couldn't come up with a good reason to have it tbh. The literal inversion would be all other nodes in the document, including non-extended footnotes and the first part of extended footnotes.

I can add it but the use case for `cr_only_if_following_inpage_footnote` is so narrow (basically just to easily apply `-cr-hint: extend-footnote-inpage`), so I don't see any use case for the negated version.

poire-z replied:
Even if no use case, let's add it, just for balance/symetry :)
(May be it can be used with some other only-if (to not match all) to counteract things? Anyway, I would try it if stuck, best to have it have some effect that just doing nothing.)

moben commented on crengine/src/lvstsheet.cpp:
```diff
@@ -5333,6 +5335,34 @@ void LVCssDeclaration::apply( css_style_rec_t * style, const ldomNode * node ) c
+                        if ( ! STYLE_HAS_CR_HINT(prevstyle, EXTEND_FOOTNOTE_INPAGE) ) {
+                            // found a node that's not part of the footnote chain
+                            return; // don't apply anything more of this declaration to this style
+                        }
```
Added a comment, let me know if it's clear now.

moben commented:
Addressed some of the initial feedback. I'll think a bit if there's a good way to un-hint hinted footnotes that don't end up being footnotes but maybe that makes sense as a separate PR? 

To confirm: There is no change in behavior if the new hint isn't used.

poire-z commented on crengine/include/cssdef.h:
```diff
@@ -465,6 +465,9 @@ enum css_generic_value_t {
 #define CSS_CR_HINT_FOOTNOTE_IGNORE         0x08000000 // -cr-hint: footnote-ignore block is not a footnote (even if
                                                        //                           everything else indicates it is)
 
+// To be set on a block element: it is a continuation of an inpage footnote and will
```
it *may be considered* a continuation...

poire-z commented on crengine/include/lvpagesplitter.h:
```diff
@@ -444,13 +444,16 @@ class LVRendPageContext
                     // document, so we can't accumulate lines from different places: override
                     // its content. (This is consistent with the way crengine handle id= when
                     // building the DOM: later ones override ealier ones).
+                    // Exception: When `allow_appending` we extend an already created footnote
+                    // trust that the caller has ensured that this is contiguous with already
+                    // accumulated lines by calling `getCurrentFootNoteId`
```
(Missing some punctuation (`,` or `:`) at end of first line)

poire-z commented on crengine/src/lvstsheet.cpp:
```diff
@@ -5366,6 +5368,36 @@ void LVCssDeclaration::apply( css_style_rec_t * style, const ldomNode * node ) c
                     }
                 }
+                else if ( only_if == cr_only_if_not_following_inpage_footnote) {
+                    if ( ! STYLE_HAS_CR_HINT(style, FOOTNOTE_INPAGE)) {
+                        ldomNode * prevsibling = node->getUnboxedPrevSibling(true);
```
You could negate the check and break here, just to have it all the indentation aligned with previous branch.

(Thought it could have been the same branch with just a check inside - and would not need this whole duplication :) but ok.)

poire-z commented on crengine/src/lvstsheet.cpp:
```diff
@@ -5366,6 +5368,36 @@ void LVCssDeclaration::apply( css_style_rec_t * style, const ldomNode * node ) c
+                        while (true) {
+                            if ( prevsibling == NULL || prevsibling->isNull() ) {
+                                break;
+                            }
+
+                            css_style_ref_t prevstyle = prevsibling->getStyle();
```
(No blank line needed here and below (also in previous branch), to keep all these condensed.)

moben commented:
Changes:
1. Reduced duplication between `(not-)following-inpage-footnote`. I know C++isms aren't the most welcome in crengine but in my opinion this is the cleanest way.
2. Added `-cr-only-if: (not-)extended-inpage-footnote`. This just matches anything that has/doesn't have `-cr-hint: extend-footnote-inpage` and was needed to fix the "smaller" footnote tweaks in https://github.com/koreader/koreader/pull/13554. If we decide there to instead add "smaller" variants for each footnote extension tweak we could drop this again.

I think I've addressed all feedback here, let me know if I missed anything or if something is still wrong/unclear.

The only thing "missing" would be re-styling nodes hinted as (extended) footnote that don't end up being shown as footnotes. But I'm fresh out of ideas how to do that without more invasive changes or recomputing styles a lot. My vote is with

> Anyway, I'm all for keeping it simple, even if it has side effects. We'll see if we need to do something better when we meet a real case where the current behaviour is not welcome, and we can't find a workaround.

poire-z commented on crengine/src/lvstsheet.cpp:
```diff
@@ -5338,65 +5338,41 @@ void LVCssDeclaration::apply( css_style_rec_t * style, const ldomNode * node ) c
-                        if ( prevstyle.isNull() ) {
-                            return; // don't apply anything more of this declaration to this style
+                else if ( only_if == cr_only_if_following_inpage_footnote || cr_only_if_not_following_inpage_footnote ) {
+                    bool does_follow_footnote = [&]() {
```
What's that ?
I haven't seen that construct in crengine yet.
(Feels like an anonymous function ? How costly it is ? Any alternative ?)

moben replied:
exactly, it's a lambda / anonymous function. In this case it's specifically an anonymous function that's immediately invoked (note the `()` at the end), so you just get the return value. In C++ land they call this an immediately invoked lambda.

There's one other lambda in crengine that I could find, though not immediately invoked: https://github.com/koreader/crengine/blob/master/crengine/src/lvstsheet.cpp#L6907-L6909

Cost-wise I'd expect the compiler to inline this 100% of the time and get rid of the function call. Then it's essentially free. It's of course not guaranteed but it's a pretty common pattern and it's a function that is only called in one place and not visible elsewhere. Though of course there's no guarantee that it's optimized that way, but only because compilers never guarantee that sort of thing.

Alternatives:
- duplicate the code for the `following` / `non-following` cases
- do something like this for each of the branches:
```
if ( only_if == cr_only_if_following_inpage_footnote )
    break;
else
    return;
```
- static function to reuse code (essentially exactly like the lambda but named function. Should also be inlined)

I think this version is the most readable but happy to change to one of the others if you'd prefer.

moben commented on crengine/src/lvstsheet.cpp:
```diff
@@ -5332,6 +5342,55 @@ void LVCssDeclaration::apply( css_style_rec_t * style, const ldomNode * node ) c
+                            return; // don't apply anything more of this declaration to this style
+                    }
+                }
+                else if ( only_if == cr_only_if_following_inpage_footnote || cr_only_if_not_following_inpage_footnote ) {
```
Thanks, good catch :+1:

moben commented on crengine/src/lvrend.cpp:
```diff
@@ -7278,14 +7278,26 @@ void renderBlockElementEnhanced( FlowState * flow, ldomNode * enode, int x, int
+            lString32 last_footnoteid = flow->getPageContext()->getCurrentFootNoteId();
+            if ( ! last_footnoteid.empty() ) {
+                enode->getAllInnerAttributeValues(attr_id, footnoteIds);
+                footnoteIds.insert(0, last_footnoteid);
```
Renamed the variable and added a comment to hopefully make it clear.

In the footnotes that are tracked in the `PageContext`, only one matters: The `actual=true` footnote that has the actual lines. Other footnote objects might be created if there are multiple Ids, but these just link to the "actual" footnote.

`getCurrentFootNoteId` gives us the `id` of this footnote (if there is one and we're allowed to extend it). So we only need that one to ensure our lines end up in the same block, and the others we pass are just to create links back to that block in case there are links to multiple different `id` anchors within the footnote. (that's all the same as it's always been with regular footnotes)

moben commented on crengine/src/lvstsheet.cpp:
```diff
@@ -5332,6 +5342,55 @@ void LVCssDeclaration::apply( css_style_rec_t * style, const ldomNode * node ) c
+                            }
+
+                            // NOTE: This needs to skip over text nodes because they don't have a style here that we can check
+                            // They can still be matched via the `autoBoxing` selector
```
I might have confused myself there or I didn't have the correct css tweak written when I added the comment, not sure. I agree it's better to drop so I did that.

I think what I meant was that `false` didn't work even though I thought it should. The `autoBoxing` node for a preceding bit of text never had the hints set when this loop checked. But I probably just had `*` in my css at that point before I learned that it doesn't match `autoBoxing`.
